src/lib.rs

```diff
@@ -1,101 +1,93 @@
-//! Simple LDAP -> Famedly Zitadel sync tool to match users between
-//! clients and our infrastructure.
-use std::path::Path;
+//! Sync tool between other sources and our infrastructure based on Zitadel.
 
-use anyhow::{bail, Context, Result};
-use ldap_poller::{ldap::EntryStatus, ldap3::SearchEntry, Cache, Ldap};
-use tokio::sync::mpsc::Receiver;
-use tokio_stream::{wrappers::ReceiverStream, StreamExt};
+use std::fmt::Debug;
+
+use anyhow::Result;
+use tracing::{error, warn};
 
 mod config;
+mod sources;
 mod user;
 mod zitadel;
 
-pub use config::{AttributeMapping, Config, FeatureFlag};
+pub use config::{Config, FeatureFlag};
+use sources::{csv::CsvSource, ldap::LdapSource, ukt::UktSource, Source};
+pub use sources::{ldap::AttributeMapping, ukt::test_helpers};
 use zitadel::Zitadel;
 
-/// Run the sync
-pub async fn sync_ldap_users_to_zitadel(config: Config) -> Result<()> {
-	if !config.feature_flags.contains(&FeatureFlag::SsoLogin) {
-		anyhow::bail!("Non-SSO configuration is currently not supported");
-	}
-
-	let cache = read_cache(&config.cache_path).await?;
-	let zitadel = Zitadel::new(&config).await?;
-	let (mut ldap_client, ldap_receiver) = Ldap::new(config.clone().ldap.into(), cache);
-
-	let deactivate_only = config.deactivate_only();
-
-	let sync_handle: tokio::task::JoinHandle<Result<_>> = tokio::spawn(async move {
-		ldap_client.sync_once(None).await.context("failed to sync/fetch data from LDAP")?;
-
-		if config.dry_run() {
-			tracing::warn!("Not writing ldap cache during a dry run");
-		} else {
-			let cache = ldap_client.persist_cache().await;
-			tokio::fs::write(
-				&config.cache_path,
-				bincode::serialize(&cache).context("failed to serialize cache")?,
-			)
-			.await
-			.context("failed to write cache")?;
-		}
-
-		tracing::info!("Finished syncing LDAP data");
+/// Sync Tool
+pub struct SyncTool {
+	/// Configuration
+	config: Config,
+	/// Sources
+	sources: Vec<Box<dyn Source + Send + Sync>>,
+}
 
-		Ok(())
-	});
+impl SyncTool {
+	/// Create a new SyncTool
+	pub fn new(config: Config) -> Result<Self> {
+		let mut sources: Vec<Box<dyn Source + Send + Sync>> = Vec::new();
 
-	let (added, changed, removed) = get_user_changes(ldap_receiver).await;
+		if config.sources.ldap.is_some() {
+			let ldap = LdapSource::new(&config)?;
+			sources.push(Box::new(ldap));
+		}
 
-	sync_handle.await??;
+		if config.sources.ukt.is_some() {
+			let ukt = UktSource::new(&config)?;
+			sources.push(Box::new(ukt));
+		}
 
-	zitadel.update_users(changed).await?;
+		if config.sources.csv.is_some() {
+			let csv = CsvSource::new(&config)?;
```

Comment:
I'd suggest 
```rust
if let Some(config) = config.sources.csv {
  let csv = CsvSource::new(&config)?;
```
to avoid unwrapping `config.sources.csv` inside `CsvSource::new`

Reply:
Done.